admin/database/postgres/postgres.go

```diff
@@ -2106,3 +2176,55 @@ func (e *wrappedError) Error() string {
 func (e *wrappedError) Unwrap() error {
 	return e.err
 }
+
+// encrypts plaintext using AES-GCM with the given key and returns the base64 encoded ciphertext
+func encrypt(plaintext string, key []byte) (string, error) {
+	block, err := aes.NewCipher(key)
+	if err != nil {
+		return "", err
+	}
+
+	aesGCM, err := cipher.NewGCM(block)
+	if err != nil {
+		return "", err
+	}
+
+	nonce := make([]byte, aesGCM.NonceSize())
+	if _, err := io.ReadFull(rand.Reader, nonce); err != nil {
+		return "", err
+	}
+
+	// nonce is prepended to the ciphertext, so it can be used for decryption
+	ciphertext := aesGCM.Seal(nonce, nonce, []byte(plaintext), nil)
+	return base64.StdEncoding.EncodeToString(ciphertext), nil
+}
+
+// decrypts the base64 encoded ciphertext using AES-GCM with the given key and returns the plaintext
+func decrypt(ciphertextBase64 string, key []byte) (string, error) {
+	ciphertext, err := base64.StdEncoding.DecodeString(ciphertextBase64)
+	if err != nil {
+		return "", err
+	}
+
+	block, err := aes.NewCipher(key)
+	if err != nil {
+		return "", err
+	}
+
+	aesGCM, err := cipher.NewGCM(block)
+	if err != nil {
+		return "", err
+	}
+
+	nonceSize := aesGCM.NonceSize()
+	if len(ciphertext) < nonceSize {
+		return "", errors.New("ciphertext too short")
+	}
+
+	nonce, ciphertext := ciphertext[:nonceSize], ciphertext[nonceSize:]
```

Comment:
Seems weird that `aesGCM.Seal` does the append automatically, but we need to manually parse it here. Is this expected?

Also, just to avoid errors, did you triple check that this is the correct way to store the nonce (e.g. that the nonce size or output size never changes)?

Reply:
```
	// Seal encrypts and authenticates plaintext, authenticates the
	// additional data and appends the result to dst, returning the updated
	// slice. The nonce must be NonceSize() bytes long and unique for all
	// time, for a given key.
	//
	// To reuse plaintext's storage for the encrypted output, use plaintext[:0]
	// as dst. Otherwise, the remaining capacity of dst must not overlap plaintext.
	Seal(dst, nonce, plaintext, additionalData []byte) []byte
```
We are deliberately using `nonce` slice as dst so that encrypted text is appended to it and no need to separately store the nonce. 
`crypto/cipher/gcm.go` ->
```
const (
	gcmBlockSize         = 16
	gcmTagSize           = 16
	gcmMinimumTagSize    = 12 // NIST SP 800-38D recommends tags with 12 or more bytes.
	gcmStandardNonceSize = 12
)
...
func NewGCM(cipher Block) (AEAD, error) {
	return newGCMWithNonceAndTagSize(cipher, gcmStandardNonceSize, gcmTagSize)
}
```
So I think nonce size is also not a problem. About storing nonce along with encrypted text, I am not an expert but I found this to be a common practice - https://crypto.stackexchange.com/questions/57895/would-it-be-safe-to-store-gcm-nonce-in-the-encrypted-output. Let me know if you have any other opinion.

Reply:
Upon looking more, seems like its not a global standard and libraries are free to choose the size. 12 is the common size but go implementation can choose to change the value of const `gcmStandardNonceSize` in newer version, not sure.

Reply:
Should we store nonce size ?

Reply:
If in newer versions go lib changes the nonce size then as per current code its not going to work anyways because they have 
```
if len(nonce) != g.nonceSize {
		panic("crypto/cipher: incorrect nonce length given to GCM")
	}
```
or they will have to make it backwards compatible somehow, so I doubt they will change it. Only use case is we use another implementation in future but there should be someway there as well to make it work with this size.

Reply:
We're discussing this offline and will follow up here